vllm/entrypoints/chat_utils.py

```diff
@@ -153,6 +155,190 @@ class ConversationMessage(TypedDict, total=False):
+    if isinstance(node, jinja2.nodes.Name):
+        return node.ctx == "load" and node.name == varname
+
+    return False
+
+
+def _is_attr_access(node: jinja2.nodes.Node, varname: str, key: str) -> bool:
+    if isinstance(node, jinja2.nodes.Getitem):
+        return (node.ctx == "load" and _is_var_access(node.node, varname)
+                and isinstance(node.arg, jinja2.nodes.Const)
+                and node.arg.value == key)
+
+    if isinstance(node, jinja2.nodes.Getattr):
+        return (node.ctx == "load" and _is_var_access(node.node, varname)
+                and node.attr == key)
+
+    return False
+
+
+def _is_var_or_elems_access(
+    node: jinja2.nodes.Node,
+    varname: str,
+    key: Optional[str] = None,
+) -> bool:
+    if isinstance(node, jinja2.nodes.Filter):
+        return (node.node is not None
+                and _is_var_or_elems_access(node.node, varname, key))
+    if isinstance(node, jinja2.nodes.Test):
+        return _is_var_or_elems_access(node.node, varname, key)
+
+    if (isinstance(node, jinja2.nodes.Getitem)
+            and isinstance(node.arg, jinja2.nodes.Slice)):
+        return _is_var_or_elems_access(node.node, varname, key)
+
+    # yapf: disable
+    return (
+        _is_attr_access(node, varname, key) if key
+        else _is_var_access(node, varname)
+    ) # yapf: enable
+
+
+def _iter_nodes_assign_var_or_elems(root: jinja2.nodes.Node, varname: str):
+    yield root, varname
+
+    for assign_ast in root.find_all(jinja2.nodes.Assign):
+        lhs = assign_ast.target
+        rhs = assign_ast.node
+
+        if _is_var_or_elems_access(rhs, varname):
+            assert isinstance(lhs, jinja2.nodes.Name)
+            yield assign_ast, lhs.name
+
+
+# NOTE: The proper way to handle this is to build a CFG so that we can handle
+# the scope in which each variable is defined, but that is too complicated
+def _iter_nodes_assign_messages_item(root: jinja2.nodes.Node):
+    messages_varnames = [
+        varname
+        for _, varname in _iter_nodes_assign_var_or_elems(root, "messages")
+    ]
+
+    # Search for {%- for message in messages -%} loops
+    for loop_ast in root.find_all(jinja2.nodes.For):
+        loop_iter = loop_ast.iter
+        loop_target = loop_ast.target
+
+        for varname in messages_varnames:
+            if _is_var_or_elems_access(loop_iter, varname):
+                assert isinstance(loop_target, jinja2.nodes.Name)
+                yield loop_ast, loop_target.name
+                break
+
+
+def _iter_nodes_assign_content_item(root: jinja2.nodes.Node):
+    message_varnames = [
+        varname for _, varname in _iter_nodes_assign_messages_item(root)
+    ]
+
+    # Search for {%- for content in message['content'] -%} loops
+    for loop_ast in root.find_all(jinja2.nodes.For):
+        loop_iter = loop_ast.iter
+        loop_target = loop_ast.target
+
+        for varname in message_varnames:
+            if _is_var_or_elems_access(loop_iter, varname, "content"):
```

Comment:
Does this also handle cases where content is reassingned?

Pseudo code example:
```
for message in messages:
    content = message["content"]
    for c in content:
        do_stuff(c)
```

Reply:
No, currently it doesn't do that. Let me think a bit about how to handle this...

Reply:
I wrote some code to enable this, but found that this causes false positives. In particular, [tool_chat_template_mistral.jinja](https://github.com/vllm-project/vllm/blob/main/examples/tool_chat_template_mistral.jinja) is detected as having OpenAI format because of L54 and L57 in the chat template. It would be quite complicated to condition the content format based on `message["role"]`...

Reply:
If we want to consider this, I think we should just build a CFG, otherwise our code would be quite unmaintainable 😅

Reply:
For future reference, here's the code I changed to enable this:

```diff
diff --git a/vllm/entrypoints/chat_utils.py b/vllm/entrypoints/chat_utils.py
index d6ab3c04e..c0edb7c24 100644
--- a/vllm/entrypoints/chat_utils.py
+++ b/vllm/entrypoints/chat_utils.py
@@ -204,21 +204,47 @@ def _is_var_or_elems_access(
     ) # yapf: enable
 
 
-def _iter_nodes_assign_var_or_elems(root: jinja2.nodes.Node, varname: str):
-    # Global variable that is implicitly defined at the root
-    yield root, varname
+def _iter_nodes_assign_var_or_elems(
+    root: jinja2.nodes.Node,
+    varname: str,
+    key: Optional[str] = None,
+):
+    if key is None:
+        # Global variable that is implicitly defined at the root
+        yield root, varname
 
     related_varnames: List[str] = [varname]
     for assign_ast in root.find_all(jinja2.nodes.Assign):
         lhs = assign_ast.target
         rhs = assign_ast.node
 
-        if any(_is_var_or_elems_access(rhs, name) for name in related_varnames):
+        if any(_is_var_or_elems_access(rhs, related_varname, key)
+               for related_varname in related_varnames):
             assert isinstance(lhs, jinja2.nodes.Name)
             yield assign_ast, lhs.name
             related_varnames.append(lhs.name)
 
 
+def _iter_nodes_assign_elem(
+    root: jinja2.nodes.Node,
+    varname: str,
+    key: Optional[str] = None,
+):
+    for loop_ast in root.find_all(jinja2.nodes.For):
+        loop_iter = loop_ast.iter
+        loop_target = loop_ast.target
+
+        if _is_var_or_elems_access(loop_iter, varname, key):
+            assert isinstance(loop_target, jinja2.nodes.Name)
+            yield loop_ast, loop_target.name
+            break
+
+    if key is not None:
+        for _, related_varname in _iter_nodes_assign_var_or_elems(
+            root, varname, key):
+            yield from _iter_nodes_assign_elem(root, related_varname)
+
+
 # NOTE: The proper way to handle this is to build a CFG so that we can handle
 # the scope in which each variable is defined, but that is too complicated
 def _iter_nodes_assign_messages_item(root: jinja2.nodes.Node):
@@ -227,16 +253,8 @@ def _iter_nodes_assign_messages_item(root: jinja2.nodes.Node):
         for _, varname in _iter_nodes_assign_var_or_elems(root, "messages")
     ]
 
-    # Search for {%- for message in messages -%} loops
-    for loop_ast in root.find_all(jinja2.nodes.For):
-        loop_iter = loop_ast.iter
-        loop_target = loop_ast.target
-
-        for varname in messages_varnames:
-            if _is_var_or_elems_access(loop_iter, varname):
-                assert isinstance(loop_target, jinja2.nodes.Name)
-                yield loop_ast, loop_target.name
-                break
+    for messages_varname in messages_varnames:
+        yield from _iter_nodes_assign_elem(root, messages_varname)
 
 
 def _iter_nodes_assign_content_item(root: jinja2.nodes.Node):
@@ -244,16 +262,8 @@ def _iter_nodes_assign_content_item(root: jinja2.nodes.Node):
         varname for _, varname in _iter_nodes_assign_messages_item(root)
     ]
 
-    # Search for {%- for content in message['content'] -%} loops
-    for loop_ast in root.find_all(jinja2.nodes.For):
-        loop_iter = loop_ast.iter
-        loop_target = loop_ast.target
-
-        for varname in message_varnames:
-            if _is_var_or_elems_access(loop_iter, varname, "content"):
-                assert isinstance(loop_target, jinja2.nodes.Name)
-                yield loop_ast, loop_target.name
-                break
+    for message_varname in message_varnames:
+        yield from _iter_nodes_assign_elem(root, message_varname, "content")
 
 
 def _try_extract_ast(chat_template: str) -> Optional[jinja2.nodes.Template]:
```